grisp/grisp_base/drivers/grisp_ir_drv.c

```diff
@@ -0,0 +1,206 @@
+    return (ErlDrvData) data;
+}
+
+void grisp_ir_output(ErlDrvData arg, char *buf, ErlDrvSizeT len){
+  struct grisp_ir_data *data = (struct grisp_ir_data *) arg;
+
+  uint8_t res, index, cmd, type, attribute;
+  rtems_name taskname;
+  struct grisp_rtems_event_arg *earg;
+  
+  index = *buf++;
+  cmd = *buf++;
+  
+  if (index >= N_PINS) {
+    driver_failure_atom(data->port, "pin_index_range");
+    return;
+  }
+  if (pins[index].pio == PIOC){
+    driver_failure_atom(data->port, "no_interrupt_allowed");
+    return;
+  }
+
+  switch(cmd){
+  case GRISP_REGISTER_IR:
+    type = *buf++;
+    attribute = *buf++;
+    
+    /* Setup Pins for Interrupt */
+    pins[index].type = type;
+    pins[index].attribute = attribute;
+
+    earg = (struct grisp_rtems_event_arg *)driver_alloc(sizeof(struct grisp_rtems_event_arg));
+    earg->data = data;
+    earg->index = index;
+    /* Start task to receive interrupts */
+    taskname = rtems_build_name('G','I','R',(char) index);
+    assert(rtems_task_create(taskname, 8, 1024, RTEMS_DEFAULT_MODES, RTEMS_DEFAULT_ATTRIBUTES, &(data->tid[index])) == RTEMS_SUCCESSFUL);
+    assert(rtems_task_start(data->tid[index], &grisp_ir_pin_trigger, (rtems_task_argument) earg) == RTEMS_SUCCESSFUL);
+    PIO_ConfigureIt(pins+index, &grisp_ir_handle, (void *) &(data->tid[index]));
+    PIO_EnableIt(pins+index);
+    driver_output(data->port, &res, 0);
+    break;
+  case GRISP_DISABLE_IR:
+    PIO_DisableIt(pins+index);
+    rtems_event_send(data->tid[index], RTEMS_EVENT_6);
+    driver_output(data->port, &res, 0);
+    break;
+  default:
+    driver_failure_atom(data->port, "unknown_command");
+    return;
+  }
+}
+
+/* RTEMS Task to receive event and write to pipe */
+rtems_task grisp_ir_pin_trigger(rtems_task_argument arg){
+  struct grisp_rtems_event_arg *earg = (struct grisp_rtems_event_arg *) arg;
+  rtems_event_set eset;
+  rtems_id tid = earg->data->tid[earg->index];
+  
+  while(rtems_event_receive(RTEMS_ALL_EVENTS, RTEMS_EVENT_ANY| RTEMS_WAIT, RTEMS_NO_TIMEOUT, &eset) == RTEMS_SUCCESSFUL){
+    if (eset == RTEMS_EVENT_5){
+      write(earg->data->fd[1], (void *) &(earg->index), sizeof(char));
+    }else if(eset == RTEMS_EVENT_6){
+      driver_free(earg);
+      rtems_task_delete(tid);
+      return;
+    }
+  }
+}
+
+/* ISR */
+void grisp_ir_handle(const Pin *pPIn, void *arg){
+    rtems_id *tid = (rtems_id *) arg;
+    rtems_event_send(*tid, RTEMS_EVENT_5);
+}
+
+/* Erlang select callback */
+void grisp_ir_ready_input(ErlDrvData arg, ErlDrvEvent event){
+    struct grisp_ir_data *data = (struct grisp_ir_data *) arg;
+    ssize_t len;
+
+    len = read(data->fd[0], (void *) &(data->buf), sizeof(char));
+    if (len<0)
+      driver_failure_posix(data->port, errno);
+    else
+      driver_output(data->port, (char *) &(data->buf), len);
+}
+
+void grisp_ir_stop(ErlDrvData arg){
+    struct grisp_ir_data *data = (struct grisp_ir_data *) arg;
+    PIO_DisableIt(triggerpin);
```

Comment:
Shouldn't we  use the `pins` array to disable all handlers ? What is `triggerpin` here ?